src/buffer/out/sgrStack.hpp

```diff
@@ -0,0 +1,238 @@
+                    }
+#pragma warning(pop)
+                }
+            }
+
+            return currentAttributes;
+        }
+
+        // Xterm allows the save stack to go ten deep, so we'll follow suit. Pushes after
+        // ten deep will still remain "balanced"--once you pop back down below ten, you'll
+        // restore the appropriate text attributes. However, if you get more than a
+        // hundred pushes deep, we'll stop counting. Why unbalance somebody doing so many
+        // pushes? Putting a bound on it allows us to provide "reset" functionality: at
+        // any given point, you can execute 101 pops and know that you've taken the stack
+        // (push count) to zero. (Then you reset text attributes, and your state is
+        // clean.)
+        static constexpr int c_MaxStoredSgrPushes = 10;
+        static constexpr int c_MaxBalancedPushes = 100;
+
+    private:
+        static constexpr uint32_t _GraphicsOptionToFlag(DispatchTypes::GraphicsOptions option)
+        {
+            int iOption = static_cast<int>(option);
+
+            if (iOption < (sizeof(uint32_t) * 8))
+            {
+                iOption = 1 << iOption;
+            }
+            // else it's a bad parameter; we'll just ignore it
+
+            return iOption;
+        }
+
+        TextAttribute _CombineWithCurrentAttributes(const TextAttribute& currentAttributes,
+                                                    const TextAttribute& savedAttribute,
+                                                    uint32_t validParts) noexcept // of savedAttribute
+        {
+            TextAttribute result = currentAttributes;
+
+            // From xterm documentation:
+            //
+            //  CSI # {
+            //  CSI Ps ; Ps # {
+            //            Push video attributes onto stack (XTPUSHSGR), xterm.  The
+            //            optional parameters correspond to the SGR encoding for video
+            //            attributes, except for colors (which do not have a unique SGR
+            //            code):
+            //              Ps = 1  -> Bold.
+            //              Ps = 2  -> Faint.
+            //              Ps = 3  -> Italicized.
+            //              Ps = 4  -> Underlined.
+            //              Ps = 5  -> Blink.
+            //              Ps = 7  -> Inverse.
+            //              Ps = 8  -> Invisible.
+            //              Ps = 9  -> Crossed-out characters.
+            //              Ps = 1 0  -> Foreground color.
+            //              Ps = 1 1  -> Background color.
+            //              Ps = 2 1  -> Doubly-underlined.
+            //
+            //  (some closing braces for people with editors that get thrown off without them: }})
+            //
+            // Attributes that are not currently supported are simply ignored.
+
+            if (_GraphicsOptionToFlag(DispatchTypes::GraphicsOptions::BoldBright) & validParts)
```

Comment:
> `_GraphicsOptionToFlag(DispatchTypes::GraphicsOptions::BoldBright) & validParts`

We're doing this style statement a whole bunch - maybe there should be another helper `IsGraphicsFlagSet(GraphicsOptions opt, uint32_t flags)`

Reply:
also alleviated by using `std::bitset::test`